atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java

```diff
@@ -629,28 +667,20 @@ private ColumnSelection getColumnSelectionForRange(RangeRequest range) {
             }
         }
 
-        ColumnSelection columns = range.getColumnNames().isEmpty()
+        ColumnSelection cols = range.getColumnNames().isEmpty()
                 ? ColumnSelection.all() : ColumnSelection.create(range.getColumnNames());
 
-        ResultWithToken results = getTimestampsByCell(table, rows, columns, timestamp, batchSize, range.isReverse());
+        ResultWithToken result = getTimestampsByCell(table, rows, cols, timestamp, batchSize, range.isReverse(), token);
 
         NavigableMap<byte[], SortedMap<byte[], Set<Long>>> cellsByRow =
-                Cells.breakCellsUpByRow(Multimaps.asMap(results.results));
-
-        List<RowResult<Set<Long>>> finalResults = Lists.newArrayListWithCapacity(results.size());
-        for (Entry<byte[], SortedMap<byte[], Set<Long>>> entry : cellsByRow.entrySet()) {
-            finalResults.add(RowResult.create(entry.getKey(), entry.getValue()));
-        }
-        byte[] nextRow = results.currentRow;
-        boolean mayHaveMoreResults = !results.finishedRange;
-        if (results.isEmpty() && nextRow != null) {
-            if (!RangeRequests.isTerminalRow(range.isReverse(), nextRow)) {
-                mayHaveMoreResults = false;
-            } else {
-                nextRow = RangeRequests.getNextStartRow(range.isReverse(), nextRow);
-            }
+                Cells.breakCellsUpByRow(Multimaps.asMap(result.entries));
+        if (range.isReverse()) {
+            cellsByRow = cellsByRow.descendingMap();
         }
-        return SimpleTokenBackedResultsPage.create(nextRow, finalResults, mayHaveMoreResults);
+        List<RowResult<Set<Long>>> finalResults = new ArrayList<>();
+        cellsByRow.entrySet().forEach(entry -> finalResults.add(RowResult.create(entry.getKey(), entry.getValue())));
```

Comment:
You can do something like:
`List<...> finalResults = cellsByRow.entrySet().forEach(entry -> RowResult.create(...)).collect(Collectors.toList())`.